org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/rename/RenameService2.xtend

```diff
@@ -83,25 +114,96 @@ class RenameService2 implements IRenameService2 {
 
 	protected def EObject getElementAtOffset(XtextResource xtextResource, Document document, Position caretPosition) {
 		val caretOffset = document.getOffSet(caretPosition)
-		return xtextResource.getElementWithIdentifierAt(caretOffset) 
+		return xtextResource.getElementWithIdentifierAt(caretOffset)
 			?: xtextResource.getElementWithIdentifierAt(caretOffset - 1)
 	}
-	
+
 	protected def getElementWithIdentifierAt(XtextResource xtextResource, int offset) {
 		if (offset >= 0) {
 			val rootNode = xtextResource?.parseResult?.rootNode
 			if (rootNode !== null) {
 				val leaf = NodeModelUtils.findLeafNodeAtOffset(rootNode, offset)
 				if (leaf !== null && leaf.isIdentifier) {
-					return xtextResource.resolveElementAt(offset)	
+					return xtextResource.resolveElementAt(offset)
 				}
 			}
 		}
 		return null
 	}
-	
+
 	protected def isIdentifier(ILeafNode leafNode) {
 		return (leafNode.grammarElement instanceof RuleCall || leafNode.grammarElement instanceof CrossReference)
-			&& !tokenUtil.isWhitespaceOrCommentNode(leafNode) 
+			&& !tokenUtil.isWhitespaceOrCommentNode(leafNode)
+	}
+
+	override prepareRename(PrepareRenameOptions options) {
+		val uri = options.params.textDocument.uri
+		options.languageServerAccess.doRead(uri) [ context |
```

Comment:
When called from `rename`, should we make sure to use a single `doRead()` transaction? The builder could kick in and change the model in between.

Reply:
- If it's already fatal, don't add another error (it's likely a follow up)
- If not, and the element could not be detected (null) or is a unresolvable line (proxy, guarded by the former check), put accept the error
Am I missing something?

Reply:
OK, it makes sense; but overriding clients can still mess it up easily. I will do it in one `doRead` for the default case.

Reply:
[Done](https://github.com/eclipse/xtext-core/pull/1051/files#diff-f0264dd6c439d644cd998f039eba6d21R81).